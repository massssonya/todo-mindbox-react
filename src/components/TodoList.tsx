import { Fragment } from "react";
import { TodoElement } from "./TodoElement";
import { ITodo } from "../types";
import {
	selectAllTodos,
	selectActiveTodos,
	selectCompletedTodos,
	selectSelector
} from "../redux/slices/todosSlice";
import { useAppSelector } from "../redux/hooks";

const selectorsDict: { [key: string]: string } = {
	selectAllTodos: "",
	selectActiveTodos: "Active",
	selectCompletedTodos: "Completed"
};

export const TodoList = () => {
	const selector = useAppSelector(selectSelector);
	const todos = useAppSelector((state) => {
		switch (selector) {
			case "selectAllTodos":
				return selectAllTodos(state);
			case "selectActiveTodos":
				return selectActiveTodos(state);
			case "selectCompletedTodos":
				return selectCompletedTodos(state);
			default:
				return [];
		}
	});

	return (
		<div className="relative flex flex-col w-full h-[250px] overflow-y-scroll">
			{todos?.length === 0 && (
				<p className="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 text-2xl text-center">
					No {selectorsDict[selector]} todos
				</p>
			)}
			{todos?.map((todo: ITodo) => (
				<Fragment key={todo.id}>
					<TodoElement todo={todo} />
				</Fragment>
			))}
		</div>
	);
};
