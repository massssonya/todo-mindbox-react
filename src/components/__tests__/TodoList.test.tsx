import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import * as reduxHooks from "../../redux/hooks";
import { TodoList } from "../TodoList";

const initialState = [
	{ id: "1", value: "test", isCompleted: false },
	{ id: "2", value: "test2", isCompleted: true }
];

jest.mock("../../redux/hooks");

describe("TodoList", () => {
	const useAppSelector = jest.spyOn(reduxHooks, "useAppSelector");

	it("should render correctly", () => {
		useAppSelector.mockReturnValue(initialState);
		render(<TodoList />);
		expect(screen.getByText("test")).toBeInTheDocument();
		expect(document.querySelectorAll("input[type='checkbox']")).not.toHaveLength(0);
	});

	it("should render no todos", () => {
		useAppSelector.mockReturnValue([]);
		render(<TodoList />);
		expect(screen.getByText("No todos")).toBeInTheDocument();
	});

	it("should render todos", () => {
		useAppSelector.mockReturnValue(initialState);
		const { container } = render(<TodoList />);
		expect(container).toMatchSnapshot();
	});
});
