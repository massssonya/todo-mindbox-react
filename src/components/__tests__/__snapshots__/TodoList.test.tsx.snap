// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should render todos 1`] = `
<div>
  <div
    class="relative flex flex-col w-full h-[250px] overflow-y-scroll"
  >
    <div
      class="w-full py-4 pl-2 flex items-center border "
    >
      <input
        class="scale-[1.5] ml-1"
        type="checkbox"
      />
      <label
        class="ml-4 p-2"
      >
        test
      </label>
    </div>
    <div
      class="w-full py-4 pl-2 flex items-center border text-gray-400 line-through"
    >
      <input
        checked=""
        class="scale-[1.5] ml-1"
        type="checkbox"
      />
      <label
        class="ml-4 p-2"
      >
        test2
      </label>
    </div>
  </div>
</div>
`;
